### PR TITLE
Add new config and lib loader

### PR DESCRIPTION
**Pull request by imbstack**
This modernizes most of the things I can see in here. It's in preparation for adding sentry when I get around to it.

Is it possible for me to deploy a testing version of this service to see if it runs on heroku properly and everything? This seems like a very big change to deploy blind.

**This will require env vars to be updated in heroku before deploy**

### CODE REVIEW

**Comment by imbstack** [comment]
Oh, we'll totally need to migrate whatever creds were in ```taskcluster-index.conf.json.enc``` to travis env variables as well. We should talk at some point and figure out what the creds were.

File: bin/server.js [hunk 1]
```diff
@@ -5,124 +5,151 @@ var debug       = require('debug')('index:bin:server');
...
-    aws:              cfg.get('aws')
-  }).then(function(validator_) {
-    validator = validator_;
-  });
+  validator: {
+    requires: ['cfg'],
+    setup: ({cfg}) => base.validator({
+      folder:           path.join(__dirname, '..', 'schemas'),
+      constants:        require('../schemas/constants'),
+      publish:          cfg.index.publishMetaData === 'true',
```
Comment (jonasfj) [inline bin/server.js]: Let's use `cfg.index.publishMetaData` directly, we have booleans with `typed-env-config`... will add comment in config.yml

File: config.yml [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+
+    # Name of task-graph table in azure table storage
+    indexedTaskTableName:         'IndexedTasks'
+    namespaceTableName:           'Namespaces'
+
+    # Azure account
+    azureAccount:                 !env AZURE_ACCOUNT
+
+    # Publish references and schemas
+    publishMetaData:              'false'
```
Comment (jonasfj) [inline config.yml]: Let's use `!env:bool PUBLISH_METADATA`
Then we don't have to do `cfg.index.publishMetaData === 'true'`

See: https://github.com/taskcluster/taskcluster-queue/blob/2ab0317576d9b7c6ca9b011a76ff36aaf7eaff3e/config.yml#L3

File: bin/server.js [hunk 3]
```diff
@@ -5,124 +5,151 @@ var debug       = require('debug')('index:bin:server');
...
+  },
 
-    // Mount API router
-    app.use('/v1', router);
+  server: {
+    requires: ['cfg', 'api'],
+    setup: async ({cfg, api}) => {
+      // Create app
+      let app = base.app({
+        port:           Number(process.env.PORT || cfg.server.port),
```
Comment (jonasfj) [inline bin/server.js]: See what we do in queue, it's smart we can use `!env:number PORT` in config.yml

File: config.yml [hunk 4]
```diff
@@ -0,0 +1,112 @@
+defaults:
+  index:
```
Comment (jonasfj) [inline config.yml]: let's call this section `app` like we do in queue...

Then elements of loader thing is easier to copy/paste between components... it's more generic..

**Comment by jonasfj** [comment]
this looks good... Fix comments and we're ready to go...

IMO this isn't many changes... most of them are in tests... and they don't break production :)

**Comment by imbstack** [comment]
Ok, that finally worked. If you want to give this one more look over, I'll give you a bit of time. Otherwise I'll ship this shortly. I've already set Heroku up with the newest env vars.

**Pull request merged**